tracker/gioui/numericupdown.go

```diff
@@ -48,14 +48,20 @@ type NumericUpDownStyle struct {
 	Tooltip         component.Tooltip
 	Width           unit.Dp
 	Height          unit.Dp
+	Padding         unit.Dp
 	shaper          text.Shaper
+	Hidden          bool
 }
 
 func NewNumberInput(v tracker.Int) *NumberInput {
 	return &NumberInput{Int: v}
 }
 
 func NumericUpDown(th *material.Theme, number *NumberInput, tooltip string) NumericUpDownStyle {
+	return NumericUpDownPadded(th, number, tooltip, 0)
+}
+
+func NumericUpDownPadded(th *material.Theme, number *NumberInput, tooltip string, padding int) NumericUpDownStyle {
```

Comment:
I think making many versions of a NumericUpdown* that return slightly different NumericUpDownStyle:s smells a bit to me. Can you just not do:

```
s := NumericUpDown(...)
s.Padding = (what you want)
```

In the case you need to change the default padding is not suitable? Or will there be so many places where we need to have something else than the default padding?

Reply:
About Hidden: the idea originated from note_editor.go layoutButtons, which is strucutred as

```
	return Surface{Gray: 37, Focus: te.scrollTable.Focused() || te.scrollTable.ChildFocused(), FitSize: true}.Layout(gtx, func(gtx C) D {
		addSemitoneBtnStyle := ActionButton(gtx, t.Theme, te.AddSemitoneBtn, "+1")
		subtractSemitoneBtnStyle := ActionButton(gtx, t.Theme, te.SubtractSemitoneBtn, "-1")
		addOctaveBtnStyle := ActionButton(gtx, t.Theme, te.AddOctaveBtn, "+12")
		subtractOctaveBtnStyle := ActionButton(gtx, t.Theme, te.SubtractOctaveBtn, "-12")
		noteOffBtnStyle := ActionButton(gtx, t.Theme, te.NoteOffBtn, "Note Off")
		deleteTrackBtnStyle := ActionIcon(gtx, t.Theme, te.DeleteTrackBtn, icons.ActionDelete, te.deleteTrackHint)
		splitTrackBtnStyle := ActionIcon(gtx, t.Theme, te.SplitTrackBtn, icons.CommunicationCallSplit, te.splitTrackHint)
		newTrackBtnStyle := ActionIcon(gtx, t.Theme, te.NewTrackBtn, icons.ContentAdd, te.addTrackHint)
		voiceUpDown := NumericUpDownPadded(t.Theme, te.TrackVoices, "Number of voices for this track", 1)
		effectBtnStyle := ToggleButton(gtx, t.Theme, te.EffectBtn, "Hex")
		uniqueBtnStyle := ToggleIcon(gtx, t.Theme, te.UniqueBtn, icons.ToggleStarBorder, icons.ToggleStar, te.uniqueOffTip, te.uniqueOnTip)
		midiInBtnStyle := ToggleButton(gtx, t.Theme, te.TrackMidiInBtn, "MIDI")
		return layout.Flex{Axis: layout.Horizontal, Alignment: layout.Middle}.Layout(gtx,
			layout.Rigid(func(gtx C) D { return layout.Dimensions{Size: image.Pt(gtx.Dp(unit.Dp(12)), 0)} }),
			layout.Rigid(addSemitoneBtnStyle.Layout),
			layout.Rigid(subtractSemitoneBtnStyle.Layout),
			layout.Rigid(addOctaveBtnStyle.Layout),
			layout.Rigid(subtractOctaveBtnStyle.Layout),
			layout.Rigid(noteOffBtnStyle.Layout),
			layout.Rigid(effectBtnStyle.Layout),
			layout.Rigid(uniqueBtnStyle.Layout),
			layout.Rigid(Label("  Voices:", white, t.Theme.Shaper)),
			layout.Rigid(voiceUpDown.Layout),
			layout.Rigid(splitTrackBtnStyle.Layout),
			layout.Flexed(1, func(gtx C) D { return layout.Dimensions{Size: gtx.Constraints.Min} }),
			layout.Rigid(midiInBtnStyle.Layout),
			layout.Flexed(1, func(gtx C) D { return layout.Dimensions{Size: gtx.Constraints.Min} }),
			layout.Rigid(deleteTrackBtnStyle.Layout),
			layout.Rigid(newTrackBtnStyle.Layout))
	})
```

and I thought a way to set midiInBtnStyle.Hidden would make as clear as code can get.

I agree that the same purpose can be fulfilled by what I found out
midiInBtnStyle = layout.Spacer{}

but setting midiInBtnStyle to EmptyWidget() wouldn't work here because the layout.Rigid(midiInBtnStyle.Layout) below still expects something that has a Layout to call.

due to missing ternary operators, I found it a bit cumbersome / didn't want too much case distinctions inside the layouting itself, but I see that the Hidden has the disadvantage that one would have to implement it for every widget again, and this is a lot of duplication.

another spontaneous idea could be:
`layout.Rigid(hiddenIf(condition, midiInBtnStyle.Layout)),`

with
```
func hiddenIf(condition bool, widget layout.Widget) layout.Widget {
  if condition {
    return layout.Spacer{}
  }
 return widget
}
```

but I'm free for any new idea

Reply:
I pushed a new version with a similar proposal, and doing so, I already included the actual hiding of the button.

this branch has the problem, that calling RTMIDIContext.InputDevices() always iterates through all available devices. The upcoming branch has a fix by caching the number of devices after looping through them once. if you want, I can also get that improvement into the current branch.